api/v1alpha1/clusterstack_webhook.go

```diff
@@ -0,0 +1,190 @@
+	if !reflect.DeepEqual(oldClusterStack.Spec.Provider, newClusterStack.Spec.Provider) {
+		allErrs = append(allErrs,
+			field.Invalid(field.NewPath("spec", "provider"), newClusterStack.Spec.Provider, "field is immutable"),
+		)
+	}
+
+	// name is immutable
+	if !reflect.DeepEqual(oldClusterStack.Spec.Name, newClusterStack.Spec.Name) {
+		allErrs = append(allErrs,
+			field.Invalid(field.NewPath("spec", "name"), newClusterStack.Spec.Name, "field is immutable"),
+		)
+	}
+
+	// channel is immutable
+	if !reflect.DeepEqual(oldClusterStack.Spec.Channel, newClusterStack.Spec.Channel) {
+		allErrs = append(allErrs,
+			field.Invalid(field.NewPath("spec", "channel"), newClusterStack.Spec.Channel, "field is immutable"),
+		)
+	}
+
+	// KubernetesVersion is immutable
+	if !reflect.DeepEqual(oldClusterStack.Spec.KubernetesVersion, newClusterStack.Spec.KubernetesVersion) {
+		allErrs = append(allErrs,
+			field.Invalid(field.NewPath("spec", "kubernetesVersion"), newClusterStack.Spec.KubernetesVersion, "field is immutable"),
+		)
+	}
+
+	return nil, aggregateObjErrors(newClusterStack.GroupVersionKind().GroupKind(), newClusterStack.Name, allErrs)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type.
+func (r *ClusterStackWebhook) ValidateDelete(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	clusterStack, ok := obj.(*ClusterStack)
+	if !ok {
+		return admission.Warnings{}, apierrors.NewBadRequest(fmt.Sprintf("expected a ClusterStack but got a %T", obj))
+	}
+
+	clusters, err := r.getClustersUsingClusterStack(ctx, clusterStack)
+	if err != nil {
+		return admission.Warnings{fmt.Sprintf("cannot validate whether clusterstack is still in use. Watch out for potentially orphaned resources: %s", err.Error())}, nil
+	}
+	if len(clusters) > 0 {
+		return admission.Warnings{}, apierrors.NewBadRequest(fmt.Sprintf("can't delete ClusterStack as there are clusters using it: [%q]", strings.Join(clusters, ", ")))
+	}
+
+	return nil, nil
+}
+
+func (r *ClusterStackWebhook) getClustersUsingClusterStack(ctx context.Context, clusterStack *ClusterStack) ([]string, error) {
+	clusterList := &clusterv1.ClusterList{}
+	if err := r.List(ctx, clusterList, &client.ListOptions{Namespace: clusterStack.Namespace}); err != nil {
+		return nil, fmt.Errorf("failed to list clusters: %w", err)
+	}
+
+	clusters := make([]string, 0, len(clusterList.Items))
+
+	// list the names of all ClusterClasses that are referenced in Cluster objects
+	for i := range clusterList.Items {
+		if clusterList.Items[i].Spec.Topology == nil {
+			continue
+		}
+
+		clusterClass := clusterList.Items[i].Spec.Topology.Class
+		if clusterClass == "" {
+			continue
+		}
+
+		cs, err := clusterstack.NewFromClusterClassProperties(clusterClass)
+		if err != nil {
+			continue
+		}
+		if clusterStack.Spec.Provider == cs.Provider && clusterStack.Spec.Name == cs.Name && clusterStack.Spec.KubernetesVersion == cs.KubernetesVersion.StringWithDot() {
```

Comment:
can you also add the channel here plz

Reply:
done via https://github.com/SovereignCloudStack/cluster-stack-operator/pull/163/commits/f9f3c38700ff161516d9c84951297825eaa66bb5